prov/util/src/util_coll.c

```diff
@@ -76,11 +77,22 @@ int ofi_av_set_union(struct fid_av_set *dst, const struct fid_av_set *src)
 				break;
 		}
 		if (j == dst_av_set->fi_addr_count) {
-			dst_av_set->fi_addr_array[temp_count++] =
-				src_av_set->fi_addr_array[i];
+			if ((temp_count+1) >= dst_av_set->max_array_size) {
+				dst_av_set->max_array_size += src_av_set->fi_addr_count;
```

Comment:
The max_array_size should be large enough that we can just fail if we ever try to go beyond the value, versus bothering to try growing the array.

But if we do bother trying to grow the array, we can't update max_array_size until after we know that the realloc has succeeded.

It would also be better to create a function to grow the array, so we can isolate that functionality, rather than duplicating it twice (3 times if you include the initial allocation).